pkg/products/threescale/reconciler.go

```diff
@@ -342,6 +392,349 @@ func (r *Reconciler) backupSystemSecrets(ctx context.Context, serverClient k8scl
+				LbEndpoints: []*envoy_api_v2_endpoint.LbEndpoint{
+					{
+						HostIdentifier: &envoy_api_v2_endpoint.LbEndpoint_Endpoint{
+							Endpoint: &envoy_api_v2_endpoint.Endpoint{
+								Address: rateLimitEndpoint,
+							}},
+					},
+				},
+			}},
+		},
+	}
+
+	// Setting up listener
+	virtualHost := v2route.VirtualHost{
+		Name:    apicastRatelimiting,
+		Domains: []string{"*"},
+
+		Routes: []*v2route.Route{
+			{
+				Match: &v2route.RouteMatch{
+					PathSpecifier: &v2route.RouteMatch_Prefix{
+						Prefix: "/",
+					},
+				},
+				Action: &v2route.Route_Route{
+					Route: &v2route.RouteAction{
+						ClusterSpecifier: &route.RouteAction_Cluster{
+							Cluster: apicastRatelimiting,
+						},
+						RateLimits: []*route.RateLimit{{
+							Stage: &wrappers.UInt32Value{Value: 0},
+							Actions: []*route.RateLimit_Action{{
+								ActionSpecifier: &route.RateLimit_Action_GenericKey_{
+									GenericKey: &route.RateLimit_Action_GenericKey{
+										DescriptorValue: "slowpath",
+									},
+								},
+							}},
+						}},
+					},
+				},
+			},
+		},
+	}
+
+	// Setting GRPC
+	clusterName := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"cluster_name": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: "ratelimit",
+				},
+			},
+		},
+	}
+
+	grpcService := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"timeout": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: "2s",
+				},
+			},
+			"envoy_grpc": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: clusterName,
+				},
+			},
+		},
+	}
+	grpcInnerService := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"grpc_service": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: grpcService,
+				},
+			},
+		},
+	}
+	httpFilterGrpc := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"domain": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: apicastRatelimiting,
+				},
+			},
+			"stage": &structpb.Value{
+				Kind: &structpb.Value_NumberValue{
+					NumberValue: 0,
+				},
+			},
+			"rate_limit_service": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: grpcInnerService,
+				},
+			},
+		},
+	}
+
+	// Setting up connection manager
+	manager := &hcm.HttpConnectionManager{
+		CodecType:  hcm.HttpConnectionManager_AUTO,
+		StatPrefix: "ingress_http",
+		RouteSpecifier: &hcm.HttpConnectionManager_RouteConfig{
+			RouteConfig: &envoyapi.RouteConfiguration{
+				Name:         "local_route",
+				VirtualHosts: []*v2route.VirtualHost{&virtualHost},
+			},
+		},
+		HttpFilters: []*hcm.HttpFilter{
+			{
+				Name:       "envoy.rate_limit",
+				ConfigType: &hcm.HttpFilter_Config{Config: httpFilterGrpc},
+			},
+			{
+				Name: "envoy.router",
+			},
+		},
+	}
+
+	pbst, err := ptypes.MarshalAny(manager)
+	if err != nil {
+		panic(err)
```

Comment:
Do we definitely want to panic here or is it sufficient to return a phase failed?

Reply:
Thanks :)